src/Feldspar/Core/Middleend/OptimizeUntyped.hs

```diff
@@ -357,6 +367,54 @@ mkLet v rhs body = mkLets ([(v,rhs)], body)
 eUnit :: UntypedFeld ValueInfo
 eUnit = aLit $ LTup []
 
+-- | Propagate type information, in particular array bounds, backwards from annotations to program inputs
+backpropTypes :: AExp -> AExp
+backpropTypes = bptop
+  where bptop (In r (App Let t [eRhs, In rb (Lambda v eBody)]))
+              = In r $ App Let t [eRhs, In rb $ Lambda v $ bptop eBody]
+        bptop (In r (Lambda v e)) = In r $ Lambda (refineType v $ bp v e) (bptop e)
```

Comment:
This looks suspiciously like it's quadratic in the number of nested lambdas. That might come back to bite you.

Reply:
Yes, I'm aware of that, but thanks for the warning anyway! I believe (famous last words) that the number of nested *top level* lambdas is quite limited. Machine generated code can of course always surprise you, though.